File: Java/src/gestorAplicacion/gestorMusica/Cancion.java
package gestorAplicacion.gestorMusica;

import java.io.Serializable;
import java.util.ArrayList;
import gestorAplicacion.gestorPersonas.Artista;

public class Cancion extends Musica implements Serializable {
	
	private static final long serialVersionUID = 1L;
	private static ArrayList<Cancion> cancionesDisponibles;
	private Artista artista;
	private Genero genero;
	private int duracion; // Duracion en segundos, para facilitar los calculos
	private int ano;
	
	static {
		cancionesDisponibles = new ArrayList<>();
	}
	
	public Cancion(String nombre, Artista artista, Genero genero, int duracion, int ano) {
		super(nombre);
		this.artista = artista;
		this.genero = genero;
		this.duracion = duracion;
		this.ano = ano;
		cancionesDisponibles.add(this);
		// Para agregar la canción al portafolio del artista
		artista.agregarCancion(this);
	}
	
	public Cancion(String nombre, Artista artista, int duracion, int ano) {
		this(nombre, artista, Genero.NO_ESPECIFICADO, duracion, ano);
	}
	
	public Artista getArtista() {
		return artista;
	}
	
	public void setArtista(Artista artista) {
		this.artista = artista;
	}
		
	public Genero getGenero() {
		return genero;
	}
	
	public void setGenero(Genero genero) {
		this.genero = genero;
	}
	
	public int getDuracion() {
		return duracion;
	}
	
	public void setDuracion(int duracion) {
		this.duracion = duracion;
	}
	
	public int getAno(){
		return ano;
	}
	public void setAno(int ano){
		this.ano = ano;
	}
	
	public static ArrayList<Cancion> getCancionesDisponibles() {
		return cancionesDisponibles;
	}
	
	public static void setCancionesDisponibles(ArrayList<Cancion> cancionesDisponibles) {
		Cancion.cancionesDisponibles = cancionesDisponibles;
	}

	@Override
	public String toString() {
		return "Se está reproduciendo la canción " + nombre;
	}
	
	public String descripcion() {
		return nombre + " - " + this.getArtista().getNombre();
	}
	
	@Override
	public void aumentarReproducciones(){
		this.reproducciones = this.reproducciones + 1; 
	}
	
	public static Cancion topCancion() {
	   	 Cancion masEscuchada = null;
	   	 int mayor = 0;
	   	 for (Cancion cancion: Cancion.getCancionesDisponibles()) {
	   		 if (cancion.getReproducciones()>mayor) {
	   			 masEscuchada=cancion;
	   			 mayor=cancion.getReproducciones();
	   		 }
	   	 }
	   	 return masEscuchada;
	   }
		
}

File: Python/src/gestorAplicacion/gestorMusica/cancion.py
from gestorAplicacion.gestorMusica.musica import Musica

class Cancion(Musica):
    __cancionesDisponibles = []
    
    def __init__(self, nombre, artista, genero, duracion, ano):
        super().__init__(nombre)
        self.__artista = artista
        self.__genero = genero
        self.__duracion = duracion
        self.__ano = ano
        Cancion.__cancionesDisponibles.append(self)
        # Para agregar la canción al portafolio del artista
        artista.agregarCancion(self)
        
    #falta simular el this    
    #falta inicializador estatico
    
    def getArtista(self):
        return self.__artista
    
    def setArtista(self, artista):
        self.__artista = artista
        
    def getGenero(self):
        return self.__genero
    
    def setGenero(self, genero):
        self.__genero = genero
    
    def getDuracion(self):
        return self.__duracion
    
    def setDuracion(self, duracion):
        self.__duracion = duracion
        
    def getAno(self):
        return self.__ano
    
    def setAno(self, ano):
        self.__ano = ano
    
    @classmethod
    def getCancionesDisponibles(cls):
        return cls.__cancionesDisponibles
    
    @classmethod
    def setCancionesDisponibles(cls, cancionesDisponibles):
        cls.__cancionesDisponibles = cancionesDisponibles
        
    def __str__(self):
        return "Se esta reproduciendo la canción " + self.getNombre()
    
    def descripcion(self):
        return "Título: " + self.getNombre() + "\nArtista: " + self.getArtista().getNombre()
    
    def aumentarReproducciones(self):
        self.setReproducciones(self.getReproducciones() + 1)

    @classmethod
    def topCancion(cls):
        masEscuchada = None
        mayor = 0
        for cancion in cls.__cancionesDisponibles:
            if cancion.getReproducciones() > mayor:
                masEscuchada = cancion
                mayor = cancion.getReproducciones()
        return masEscuchada
    
    @classmethod
    def topCancionGenero(cls, genero):
        masEscuchada = None
        mayor = 0
        for cancion in cls.__cancionesDisponibles:
            if cancion.getReproducciones() > mayor and cancion.getGenero() == genero:
                masEscuchada = cancion
                mayor = cancion.getReproducciones()
        return masEscuchada
        